third_party/move/move-compiler-v2/src/inliner.rs

```diff
@@ -0,0 +1,1186 @@
+//   - InlinedCall() with no nested Return() can be flattened.
+//   - others?
+// - do we need to insert FreezeRef in actual args if formal param is &T but arg is &mod T ?
+// - do we need to do anything about abilities?
+// - if lambda parameters also may be referred to by temporaries, then `rewrite_invoke` might need to call yet
+//   another `ExpRewriterFunctions` implementation to do that.
+
+use codespan_reporting::diagnostic::Severity;
+use itertools::chain;
+use move_model::{
+    ast::{Exp, ExpData, Operation, Pattern, TempIndex},
+    exp_rewriter::ExpRewriterFunctions,
+    model::{FunId, GlobalEnv, Loc, NodeId, Parameter, QualifiedId, TypeParameter},
+    symbol::Symbol,
+    ty::{ReferenceKind, Type},
+};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    fmt::Debug,
+    iter::{zip, IntoIterator, Iterator},
+    ops::Deref,
+    vec::Vec,
+};
+
+// ======================================================================================
+// Entry
+
+// Run inlining on current program's AST.  For each function which is target of the compilation,
+// visit that function body and inline any calls to functions marked as "inline".
+pub fn run_inlining(env: &mut GlobalEnv, debug: bool) {
+    // Start with targets.
+    let mut todo = BTreeSet::new();
+    // While we're iterating, error on any target inline functions lacking a body to inline.
+    let mut has_error: bool = false;
+    for module in env.get_modules() {
+        if module.is_target() {
+            for func in module.get_functions() {
+                let id = func.get_qualified_id();
+                if func.is_inline() {
+                    let func_def = func.get_def();
+                    if !func_def.deref().is_some() {
+                        has_error = true;
+                        let func_loc = func.get_loc();
+                        let func_name = func.get_name_str();
+                        if func.is_native() {
+                            let msg = format!("Inline function {} must not be native", func_name);
+                            env.diag(Severity::Error, &func_loc, &msg);
+                        } else {
+                            let msg = format!(
+                                "ICE: No body found for non-native inline function {}",
+                                func_name
+                            );
+                            env.diag(Severity::Bug, &func_loc, &msg);
+                        }
+                    }
+                } else {
+                    todo.insert(id);
+                }
+            }
+        }
+    }
+    if has_error {
+        return;
+    }
+    // Recursively find callees of each target with a function body.
+    let mut function_callees: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
+    let mut all_functions = BTreeSet::new();
+
+    // Record for all call sites from function a to B (for error messages)
+    let mut call_site_locations: BTreeMap<
```

Comment:
nit: I think this type is "complicated" enough and used as a function parameter type, that it might be worth introducing a type alias/new type?

Reply:
OK.  Also added an alias QualifiedFuncId = QualifiedId<FuncId>.